controllers/pkg/reconcilers/repository/reconciler.go

```diff
@@ -0,0 +1,210 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"code.gitea.io/sdk/gitea"
+	"github.com/go-logr/logr"
+	ctrlconfig "github.com/henderiw-nephio/nephio-controllers/controllers/config"
+	"github.com/henderiw-nephio/nephio-controllers/pkg/applicator"
+	infrav1alpha1 "github.com/nephio-project/api/infra/v1alpha1"
+	"github.com/nephio-project/nephio/controllers/pkg/giteaclient"
+	"github.com/nokia/k8s-ipam/pkg/meta"
+	"github.com/nokia/k8s-ipam/pkg/resource"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+/*
+func init() {
+	controllers.Register("repositories", &reconciler{})
+}
+*/
+
+const (
+	finalizer = "infra.nephio.org/finalizer"
+	// errors
+	errGetCr        = "cannot get cr"
+	errUpdateStatus = "cannot update status"
+)
+
+//+kubebuilder:rbac:groups=infra.nephio.org,resources=repositories,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infra.nephio.org,resources=repositories/status,verbs=get;update;patch
+
+// Setup sets up the controller with the Manager.
+func (r *Reconciler) Setup(mgr ctrl.Manager, cfg *ctrlconfig.ControllerConfig) (map[schema.GroupVersionKind]chan event.GenericEvent, error) {
+	if err := infrav1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
+		return nil, err
+	}
+
+	r.APIPatchingApplicator = applicator.NewAPIPatchingApplicator(mgr.GetClient())
+	r.giteaClient = cfg.GiteaClient
+	r.finalizer = resource.NewAPIFinalizer(mgr.GetClient(), finalizer)
+
+	return nil, ctrl.NewControllerManagedBy(mgr).
+		Named("RepositoryController").
+		For(&infrav1alpha1.Repository{}).
+		Complete(r)
+}
+
+type Reconciler struct {
+	applicator.APIPatchingApplicator
+	giteaClient giteaclient.GiteaClient
+	finalizer   *resource.APIFinalizer
+
+	l logr.Logger
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.l = log.FromContext(ctx)
+	r.l.Info("reconcile", "req", req)
+
+	cr := &infrav1alpha1.Repository{}
+	if err := r.Get(ctx, req.NamespacedName, cr); err != nil {
+		// if the resource no longer exists the reconcile loop is done
+		if resource.IgnoreNotFound(err) != nil {
+			r.l.Error(err, "cannot get resource")
+			return ctrl.Result{}, errors.Wrap(resource.IgnoreNotFound(err), "cannot get resource")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// check if client exists otherwise retry
+	giteaClient := r.giteaClient.Get()
+	if giteaClient == nil {
+		err := fmt.Errorf("gitea server unreachable")
+		r.l.Error(err, "cannot connect to git server")
+		cr.SetConditions(infrav1alpha1.Failed(err.Error()))
+		return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, errors.Wrap(r.Status().Update(ctx, cr), errUpdateStatus)
```

Comment:
Seems pretty aggressive to retry every 5 seconds forever. I thought controller runtime automatically implements exponential backoff (this is hearsay...)?

Reply:
yes I can just requeue.

Reply:
changed